src/lib/comms/sol-lwm2m-security.c

```diff
@@ -0,0 +1,470 @@
+    if (!obj_ctx) {
+        SOL_WRN("LWM2M Security object not provided!");
+        return -ENOENT;
+    }
+
+    if (!obj_ctx->instances.len) {
+        SOL_WRN("There are no Security Server instances");
+        return -ENOENT;
+    }
+
+    SOL_VECTOR_FOREACH_IDX (&obj_ctx->instances, instance, i) {
+        r = read_resources(lwm2m_client, obj_ctx, instance, res, 3,
+            SECURITY_SECURITY_MODE,
+            SECURITY_PUBLIC_KEY_OR_IDENTITY,
+            SECURITY_SECRET_KEY);
+        SOL_INT_CHECK_GOTO(r, < 0, err_clear);
+
+        //If -ENOENT when reading SECURITY_SECRET_KEY or
+        // SECURITY_SECURITY_MODE != SOL_LWM2M_SECURITY_MODE_PRE_SHARED_KEY
+        if (res[2].data_len == 0 ||
+            res[0].data[0].content.integer != SOL_LWM2M_SECURITY_MODE_PRE_SHARED_KEY) {
+            clear_resource_array(res, sol_util_array_size(res));
+            continue;
+        }
+
+        if (sol_str_slice_eq(sol_str_slice_from_blob(res[1].data[0].content.blob), id)) {
+            if (res[2].data[0].content.blob->size != SOL_DTLS_PSK_KEY_LEN) {
+                SOL_WRN("The PSK '%.*s' is %zu-bytes long; expecting a %d-bytes long PSK",
+                    SOL_STR_SLICE_PRINT(sol_str_slice_from_blob(res[2].data[0].content.blob)),
+                    res[2].data[0].content.blob->size, SOL_DTLS_PSK_KEY_LEN);
+                r = -EINVAL;
+                goto err_clear;
+            }
+
+            memcpy(psk, res[2].data[0].content.blob->mem, res[2].data[0].content.blob->size);
+            clear_resource_array(res, sol_util_array_size(res));
+            return (ssize_t)SOL_DTLS_PSK_KEY_LEN;
+        }
+
+        clear_resource_array(res, sol_util_array_size(res));
+    }
+
+    return -ENOENT;
+
+err_clear:
+    clear_resource_array(res, sol_util_array_size(res));
+    return r;
+}
+
+static ssize_t
+get_id_from_server_or_bs_server(const void *data, struct sol_network_link_addr *addr,
+    char *id, size_t id_len)
+{
+    //XXX: Assuming Clients' Endpoint Name will be used as PSK ID.
+    // Any UTF-8 String would do, (such as IP Addresses in dotted notation,
+    // domain names, etc) [RFC4279#section-5.1] and should be previously
+    // agreed upon between communication parties.
+    //XXX: Assuming communication with an LWM2M Client called cli1.
+    struct sol_str_slice cli_name;
+    const struct sol_lwm2m_security *ctx = data;
+    struct sol_lwm2m_server *lwm2m_server;
+    struct sol_lwm2m_bootstrap_server *lwm2m_bs_server;
+    struct sol_vector *known_psks;
+    struct sol_lwm2m_security_psk *stored_psk;
+    uint16_t i;
+
+    //If the caller expects a PSK_ID_LEN less than 16 bytes
+    if (id_len < SOL_DTLS_PSK_ID_LEN)
+        return -ENOBUFS;
+
+    if (ctx->type == LWM2M_SERVER) {
+        SOL_DBG("ctx->type = LWM2M_SERVER");
+        lwm2m_server = ctx->entity;
+        known_psks = lwm2m_server->known_psks;
+        cli_name = sol_str_slice_from_str("cli1");
```

Comment:
This is exactly what we were discussing the other day and I thought it was ruled "out of scope" since the LWM2M spec doesn't really have a say on how this is done.

Maybe we can discuss this idea of a callback offline on IRC... even if we do provide a callback, what should be its parameters? I mean, the server will retrieve a PSK ID based on what information, since at this point it doesn't even have the client name itself, since registration didn't occur yet. IP address, which is the thing we used at the client side, feels really weird to be used here, since Clients' IP addresses should be expected to change and so on.